## Тема “Обучение с учителем”

### Задание 1

Импортируйте библиотеки pandas и numpy.

Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.

Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42. 

Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.


In [104]:
import numpy as np
import pandas as pd

In [105]:
from sklearn.datasets import load_boston

In [106]:
boston = load_boston()

In [107]:
target = boston.target

In [108]:
feature_names = boston.feature_names

In [109]:
X = pd.DataFrame(data, columns=feature_names)

In [110]:
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [111]:
y = pd.DataFrame(target, columns=['price'])
y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

In [112]:
from sklearn.model_selection import train_test_split

In [113]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)

In [114]:
from sklearn.linear_model import LinearRegression

In [115]:
lr = LinearRegression()

In [116]:
lr.fit(X_train, y_train)

LinearRegression()

In [117]:
y_pred = lr.predict(X_test)

In [118]:
check_test = pd.DataFrame({'y_test': y_test['price'],
                          'y_pred': y_pred.flatten()},
                         columns=['y_test', 'y_pred'])

In [119]:
check_test.head(10)

y_test     y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280
76     20.0  23.146689
316    17.8  17.392124
140    14.0  14.078599
471    19.6  23.036927
500    16.8  20.599433

In [178]:
from sklearn.metrics import r2_score

In [121]:
mean_squared_error(y_test, y_pred)


21.517444231176903

In [179]:
r2_score(y_test, y_pred)

0.7112260057484974

### Задание №2

Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.

Сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42.

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.

Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.

Напишите в комментариях к коду, какая модель в данном случае работает лучше.


In [184]:
from sklearn.ensemble import RandomForestRegressor

In [185]:
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

In [188]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [189]:
y_pred_model = model.predict(X_test)
check_test_model = pd.DataFrame({
    "y_test": y_test["price"], 
    "y_pred_model": y_pred_model.flatten()})

check_test_model.head()

y_test  y_pred_model
173    23.6     22.806412
274    32.4     31.131464
491    13.6     16.339125
72     22.8     23.810726
452    16.1     17.139521

In [191]:
r2_score(y_test, y_pred_model)

0.87472606157312

Результат предсказания, полученный с помощью модели основанной на RandomForestRegressor, будет более точным, чем полученный с помощью LinearRegression. 

R2 для LinearRegression =  0.7112260057484974


R2 для RandomForestRegressor =  0.87472606157312

### Задание №3*

Вызовите документацию для класса RandomForestRegressor, найдите информацию об атрибуте feature_importances_.
С помощью этого атрибута найдите сумму всех показателей важности, установите, какие два признака показывают наибольшую важность.


In [194]:
? RandomForestRegressor

feature_importances_ : ndarray of shape (n_features,)
    The impurity-based feature importances.
    The higher, the more important the feature.
    The importance of a feature is computed as the (normalized)
    total reduction of the criterion brought by that feature.  It is also
    known as the Gini importance.

    Warning: impurity-based feature importances can be misleading for
    high cardinality features (many unique values). See
    :func:`sklearn.inspection.permutation_importance` as an alternative.

In [193]:
model.feature_importances_

array([0.03167574, 0.00154252, 0.00713813, 0.00123624, 0.01426897,
       0.40268179, 0.01429864, 0.06397257, 0.00528122, 0.01152493,
       0.01808108, 0.01245085, 0.41584732])

In [202]:
feature_importance = pd.DataFrame({'name':X.columns, 
                                   'feature_importance':model.feature_importances_}, 
                                  columns=['feature_importance', 'name'])
feature_importance.sort_values(by='feature_importance', ascending = [False])

feature_importance     name
12            0.415847    LSTAT
5             0.402682       RM
7             0.063973      DIS
0             0.031676     CRIM
10            0.018081  PTRATIO
6             0.014299      AGE
4             0.014269      NOX
11            0.012451        B
9             0.011525      TAX
2             0.007138    INDUS
8             0.005281      RAD
1             0.001543       ZN
3             0.001236     CHAS

In [203]:
feature_importance.nlargest(2, 'feature_importance')

feature_importance   name
12            0.415847  LSTAT
5             0.402682     RM

Признаки LSTAT и RM обладают наибольшей вайжностью.